### DISCUSSION TITLE
How to make a hsl highlighter using mini.hipatterns?

### DISCUSSION THREAD

**Discussion by craftzdog** [nvim-mini/mini.nvim#500] [Q&A] [answered]
It can easily highlight hex colors by using `MiniHipatterns.gen_highlighter.hex_color`.
But I often use colors in HSL.
Could you tell me how to make one for hsl?
My current config looks like so:

```lua
    local hi = require("mini.hipatterns")

    local hsl_pattern = "hsl.()%d+(), ()%d+(), ()%d+()"

    return {
      highlighters = {
        hex_color = hi.gen_highlighter.hex_color({ priority = 2000 }),
        hsl_color = {
          pattern = function()
            return hsl_pattern
          end,
          group = function(_, _, data)
            print(vim.inspect(data))
            return MiniHipatterns.compute_hex_color_group(data.full_match, "full")
          end,
          priority = 2000,
        },
      },
```

Any help would be appreciated!

**Comment by echasnovski** [comment, marked as answer]
Basically, this boils down to programming two parts:
- [Pattern to be highlighted](https://github.com/echasnovski/mini.nvim/blob/1fdbb864e2015eb6f501394d593630f825154385/doc/mini-hipatterns.txt#L149-L161).
- [Group used to highlight pattern matches](https://github.com/echasnovski/mini.nvim/blob/1fdbb864e2015eb6f501394d593630f825154385/doc/mini-hipatterns.txt#L163-L174). 

In case of HSL the real problem is to convert matched color representation in HSL to its hex string (which is the only big color space Neovim highlight groups support).

Here is an example of how I would do it to highlight one of the CSS formats like `hsl(50 80% 40%)` (taken from [here](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/hsl)):

```lua
local hsl_to_hex = function(h, s, l)
  -- Actually convert h, s, l numbers into hex color in '#RRGGBB' format
  return '#111111'
end

require('mini.hipatterns').setup({
  highlighters = {
    hsl_color = {
      pattern = 'hsl%(%d+ %d+%% %d+%%%)',
      group = function(_, match)
        local h, s, l = match:match('hsl%((%d+) (%d+)%% (%d+)%%%)')
        h, s, l = tonumber(h), tonumber(s), tonumber(l)
        local hex_color = hsl_to_hex(h, s, l)
        return MiniHipatterns.compute_hex_color_group(hex_color, 'bg')
      end,
    },
  },
})
```

**Reply by craftzdog** [reply]
Got it to work!! Thank you so much for your help❤️

**Reply by echasnovski** [reply]
That's great!

If that is not a big secret, would you mind sharing your final approach? What did you end up using for actual conversion?

**Comment by craftzdog** [comment]
Here is my config: https://gist.github.com/craftzdog/8012c78715414de7b5608bda08c77105

Result:

<img width="459" alt="Screenshot 2023-10-03 at 16 28 06" src="https://github.com/echasnovski/mini.nvim/assets/1332805/95e42ddf-a3a0-4e27-859d-2bb0a82eef45">


Hope it's helpful for someone :)

**Reply by echasnovski** [reply]
Thanks for sharing!

A summary (in case the gist will go away): a mine code snippet from earlier with a `hsl_to_hex()` adapted from https://github.com/EmmanuelOga/columns/blob/master/utils/color.lua.

**Comment by tigion** [comment]
Another variant like https://www.w3.org/TR/css-color-3/#hsl-color:

```lua
    -- Returns hex color group for matching hsl() color.
    --
    ---@param match string
    ---@return string
    local hsl_color = function(_, match)
      local style = 'bg' -- 'fg' or 'bg'
      local hue, saturation, lightness = match:match('hsl%((%d+), ?(%d+)%%, ?(%d+)%%%)')

      -- Converts HSL to RGB.
      -- https://www.w3.org/TR/css-color-3/#hsl-color
      --
      ---@param h string The hue value in degrees.
      ---@param s string The saturation value in percent.
      ---@param l string The lightness value in percent.
      ---@return integer, integer, integer
      local function hsl_to_rgb(h, s, l)
        h, s, l = h % 360, s / 100, l / 100
        if h < 0 then h = h + 360 end
        local function f(n)
          local k = (n + h / 30) % 12
          local a = s * math.min(l, 1 - l)
          return l - a * math.max(-1, math.min(k - 3, 9 - k, 1))
        end
        return f(0) * 255, f(8) * 255, f(4) * 255
      end

      local red, green, blue = hsl_to_rgb(hue, saturation, lightness)
      local hex = string.format('#%02x%02x%02x', red, green, blue)
      return hipatterns.compute_hex_color_group(hex, style)
    end
```

```lua
    hipatterns.setup({
      highlighters = {
        hsl_color = {
          pattern = 'hsl%(%d+, ?%d+%%, ?%d+%%%)',
          group = hsl_color,
        },
      },
    })
```
